core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -208,12 +226,40 @@ private void cleanExpiredFiles(List<Snapshot> snapshots, Set<Long> validIds, Set
       }
     }
 
+    // find manifests to clean up that were only referenced by snapshots that have expired
     Set<String> manifestListsToDelete = Sets.newHashSet();
     Set<String> manifestsToDelete = Sets.newHashSet();
     Set<ManifestFile> manifestsToRevert = Sets.newHashSet();
     for (Snapshot snapshot : base.snapshots()) {
       long snapshotId = snapshot.snapshotId();
       if (!validIds.contains(snapshotId)) {
+        // determine whether the changes in this snapshot are in the current table state
+        if (pickedAncestorSnapshotIds.contains(snapshotId)) {
+          // this snapshot was cherry-picked into the current table state, so skip cleaning it up. its changes will
+          // expire when the picked snapshot expires.
+          // A -- C (source=B)
+          //  `- B <-- this commit
+          continue;
+        }
+
+        long sourceSnapshotId = PropertyUtil.propertyAsLong(
+            snapshot.summary(), SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP, -1);
+        if (ancestorIds.contains(sourceSnapshotId)) {
+          // this commit was cherry-picked from a commit that is in the current table state. do not clean up its
+          // changes because it would revert data file additions that are in the current table.
+          // A -- B
+          //  `- C (source=B) <-- this commit
+          continue;
+        }
+
+        if (pickedAncestorSnapshotIds.contains(sourceSnapshotId)) {
+          // this commit was cherry-picked from a commit that is in the current table state. do not clean up its
+          // changes because it would revert data file additions that are in the current table.
+          // A -- C -- E (source=B)
+          //  `- B `- D (source=B) <-- this commit
```

Comment:
Having D (cherry pick of B) isn't possible, if B is already an ancestor, encountered this while writing UTs.

Reply:
The case that would cause this is:
* A is committed
* B is staged
* C is committed
* B is picked to create D
* The table is rolled back to C
* B is picked to create E

The main thing is the rollback, which can cause alternate branches with the same commit picked. I would expect in practice to have other commits (like one between C and D that is removed by the rollback) but we don't need those to test this case.